### PR TITLE
feat(580): support for annotations

### PR DESCRIPTION
**Pull request by r3rastogi**
Read in the annotations as described in screwdriver-cd/data-schema#145.

Pipeline-level annotations are separate than job-level annotations. Job-level annotations override shared annotations.

Related to screwdriver-cd/screwdriver#580.

### CODE REVIEW

**Review by tkyi — COMMENTED** [review]
What happens if a user specifies a pipeline level annotation that conflicts with a job level annotation? Or which one takes priority between pipeline level annotations and shared annotations?
Do we have test cases for those?

**Comment by r3rastogi** [comment]
Pipeline-level annotations and job-level annotations are separate entities. Only shared/job will collapse into job-level (and job-level will override shared). I've included additional tests to further illustrate this point.

File: test/data/shared-job-annotations.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+shared:
+    annotations:
+        foo: a
+        bar: b
+
+jobs:
+    main:
+        annotations:
+            foo: c
+            bar: d
```
Comment (d2lam) [inline test/data/shared-job-annotations.yaml]: Can we modify this so that the job level has `foo`, and not `bar`? So that we know it will pick up `bar` from `shared`, and not completely override.

**Comment by tkyi** [comment]
@r3rastogi Oh okay I see. So for our specific use case, we'd want to put the executor annotations at the job/shared level, never the pipeline level. But there will be other features in the future where it would make sense to put things in the pipeline level. Annotations that belong in pipeline vs shared/job level should never really overlap.

File: index.js [hunk 2]
```diff
@@ -44,10 +44,12 @@ module.exports = function configParser(yaml, templateFactory) {
         .then(phaseGeneratePermutations)
         // Output in the right format
         .then(doc => ({
+            annotations: Hoek.reach(doc, 'annotations', { default: {} }),
```
Comment (minzcmu) [inline index.js:47]: For the pipeline level annotation, I don't think we can do it now without sequelize 4.x since we cannot create a new field `annotations`. :(

When it goes through the pipeline.sync, it will only update the workflow for the pipeline.
https://github.com/screwdriver-cd/models/blob/master/lib/pipeline.js#L251

It probably won't break anything but it's also not doing anything.. we won't be able to see it from `/pipelines/id`.

**Review by stjohnjohnson — APPROVED** [review]
Nice!

File: test/data/shared-job-annotations.yaml (hunk 1 above)
Comment (r3rastogi) [reply test/data/shared-job-annotations.yaml]: Sounds good. Updated in next push.

**Comment by r3rastogi** [comment]
@tkyi yes, that's correct

File: index.js (hunk 2 above)
Comment (r3rastogi) [reply index.js:47]: If we've already included it in `data-schema`, why does it also have to be in `models`?

Comment (d2lam) [reply index.js:47]: @r3rastogi The config-parser only parsers the yaml.

For jobs, it goes through the parser, then store the environment, settings, etc. under `permutations`. And because`permutations` is just an object, we can just add additional fields without adding a new column in the database. Example: https://api.screwdriver.cd/v4/jobs/1

For pipeline, we need to add a new field, because it's not under another object. Sorry I remember telling you we don't need to. 

Basically, you don't need to modify the jobs table, but you need to modify the pipeline table. And since sequelize 4.x doesn't work yet, we can't add a new field. I think it's best to not implement pipeline annotations for now. And when we get sequelize to work, then we can add it.

Comment (r3rastogi) [reply index.js:47]: Ahh got it. Thanks for the explanations @d2lam and @minz1027 .

I think we should leave this code as-is for now, and add in the `models` in a future PR to `data-schema` when relevant. As @minz1027 said, this won't break anything. I'll add a note to the original issue (https://github.com/screwdriver-cd/screwdriver/issues/580) indicating that we still need to modify the`pipeline` database.

Comment (d2lam) [reply index.js:47]: @r3rastogi  According to @FenrirUnbound, you should be able to manually alter the table to have the field. I'm not a fan of having code that is not doing anything. It's a little confusing. I think we should either alter the table, or remove code that is not doing anything. 

@FenrirUnbound  can you help reetika update the table? :)

Comment (FenrirUnbound) [reply index.js:47]: It's a SQL table, so just add a column with a sane default value

Comment (r3rastogi) [reply index.js:47]: @FenrirUnbound How do you get to the table? Are their onboarding or other docs for that?

Comment (r3rastogi) [reply index.js:47]: Adding a new column using the following SQL command:
`ALTER TABLE "public"."pipelines" ADD COLUMN "annotations" text;`

@FenrirUnbound annotations are an optional field, so I don't think having a default value is necessary.

Comment (r3rastogi) [reply index.js:47]: Pipelines table now looks like:
image

Comment (d2lam) [reply index.js:47]: Cool. I see it in production tables. I think you also need it for `beta`

**Review by d2lam — APPROVED** [review]
LGTM after the beta table is updated.

File: index.js (hunk 2 above)
Comment (r3rastogi) [reply index.js:47]: Oops, didn't realize that was a thing. Added to `beta_pipelines`. I'm not sure what `beta_rc_pipelines` is though, so I didn't modify that one.
